src/SSHDebugPS/ConnectionManager.cs

```diff
@@ -149,5 +174,140 @@ public static Connection GetSSHConnection(string name)
+                    resetEvent.Set();
+                });
+
+                commandRunner.OutputReceived += ((sender, args) =>
+                {
+                    if (!string.IsNullOrWhiteSpace(args))
+                    {
+                        if (args.Trim()[0] != '{')
+                        {
+                            // output isn't json, command Error
+                            string errorMessage = string.Format(CultureInfo.CurrentCulture, UIResources.CommandExecutionErrorFormat, dockerCommand, args);
+                            throw new CommandFailedException(errorMessage);
+                        }
+
+                        var containerInstance = DockerContainerInstance.Create(args);
+                        if (containerInstance != null)
+                            containers.Add(containerInstance);
+                    }
+                });
+
+                commandRunner.Run();
+                resetEvent.WaitOne();
+
+                // might need to throw an exception here too??
+                if (exitCode != 0)
+                {
+                    Debug.Fail($"Exit Code: {exitCode}");
+                    return null;
+                }
+
+                return containers;
+            }
+            catch (Win32Exception ex)
+            {
+                // docker doesn't exist 
+                string errorMessage = string.Format(CultureInfo.CurrentCulture, UIResources.CommandExecutionErrorFormat, dockerCommand, ex.Message);
+                throw new CommandFailedException(errorMessage, ex);
+            }
+        }
+
+        public static IEnumerable<IContainerInstance> GetRemoteDockerContainers(IConnection connection)
+        {
+            SSHConnection sshConnection = connection as SSHConnection;
+            List<string> outputLines = new List<string>();
+            StringBuilder errorSB = new StringBuilder();
+            if (sshConnection == null)
+            {
+                return null;
+            }
+
+            List<DockerContainerInstance> containers = new List<DockerContainerInstance>();
+            RemoteCommandRunner commandRunner = new RemoteCommandRunner(dockerCommand, dockerPSArgs, sshConnection);
+
+            ManualResetEvent resetEvent = new ManualResetEvent(false);
+            int exitCode = 0;
+            commandRunner.ErrorOccured += ((sender, args) =>
+            {
+                resetEvent.Set();
+            });
+
+            commandRunner.Closed += ((sender, args) =>
+            {
+                exitCode = args;
+                resetEvent.Set();
+            });
+
+            commandRunner.OutputReceived += ((sender, args) =>
+            {
+                if (!string.IsNullOrWhiteSpace(args))
+                {
+                    // If it isn't json, assume its an error message
+                    if (args.Trim()[0] != '{')
+                    {
+                        errorSB.Append(args);
+                    }
```

Comment:
Can this event be fired more than once?

Reply:
Technically yes, in my test scenario no.

---
In reply to: [287513281](https://github.com/microsoft/MIEngine/pull/868#discussion_r287513281) [](ancestors = 287513281)

Reply:
Which means i need an output string builder and then to process the string afterwards.

---
In reply to: [287545722](https://github.com/microsoft/MIEngine/pull/868#discussion_r287545722) [](ancestors = 287545722,287513281)

Reply:
well now that i think about it, i think this is fine. liblinux returns lines so this *should* be ok

---
In reply to: [287545780](https://github.com/microsoft/MIEngine/pull/868#discussion_r287545780) [](ancestors = 287545780,287545722,287513281)